docs: update openid-connect plugin document (#7406)
apache/apisix: docs/zh/latest/plugins/openid-connect.md

title: openid-connect
keywords:
  - APISIX
  - API Gateway
  - OpenID Connect
  - OIDC

> description: OpenID Connect（OIDC）是基于 OAuth 2.0 的身份认证协议，APISIX 可以与支持该协议的身份认证服务对接，如 Okta、Keycloak、Ory、Authing 等，实现对客户端请求的身份认证。

Reviewer: ditto

> description: OpenID Connect（OIDC）是基于 OAuth 2.0 的身份认证协议，APISIX 可以与支持该协议的身份认证服务对接，如 Okta、Keycloak、Ory.sh、Authing 等，实现对客户端请求的身份认证。

Reviewer 2: ```suggestion
description: OpenID Connect（OIDC）是基于 OAuth 2.0 的身份认证协议，APISIX 可以与支持该协议的身份认证服务对接，如 Okta、Keycloak、Ory、Authing 等，实现对客户端请求的身份认证。
```

## 描述

> [OpenID Connect](https://openid.net/connect/)（OIDC）是基于 OAuth 2.0 的身份认证协议，APISIX 可以与支持该协议的身份认证服务对接，如 Okta、Keycloak、Ory、Authing 等，实现对客户端请求的身份认证。

Reviewer: ditto

| ------------------------------------ | ------- | ------ | --------------------- | ------------- | ------------------------------------------------------------------------------------------------ |
| client_id                            | string  | 是     |                       |               | OAuth 客户端 ID。                                                                                 |
| client_secret                        | string  | 是     |                       |               | OAuth 客户端 secret。                                                                            |
| discovery                            | string  | 是     |                       |               | 身份认证服务暴露的服务发现端点。                                                                            |
| scope                                | string  | 否     | "openid"              |               | 用于认证的范围。                                                                                  |
| realm                                | string  | 否     | "apisix"              |               | 与租户概念类似，不同 Realm 之间是相互隔离的，只能管理和验证它们所具有的用户。                                                                                  |

> | bearer_only                          | boolean | 否     | false                 |               | 当设置为 `true` 时，将检查请求头中的令牌（Token）。                                               |

Reviewer: ```suggestion
| bearer_only                          | boolean | 否     | false                 |               | 当设置为 `true` 时，将仅检查请求头中的令牌（Token）。                                               |
```

| logout_path                          | string  | 否     | "/logout"             |               | 登出路径。                                                                                        |
| post_logout_redirect_uri             | string  | 否     |                       |               | 调用登出接口后想要跳转的 URL。                                                                     |
| redirect_uri                         | string  | 否     | "ngx.var.request_uri" |               | 身份提供者重定向返回的 URI。                                                                       |
| timeout                              | integer | 否     | 3                     | [1,...]       | 请求超时时间，单位为秒                                                                             |
| ssl_verify                           | boolean | 否     | false                 | [true, false] | 当设置为 `true` 时，验证身份提供者的 SSL 证书。                                                     |
| introspection_endpoint               | string  | 否     |                       |               | 身份服务器的令牌认证端点。                                                                    |

> | introspection_endpoint_auth_method   | string  | 否     | "client_secret_basic" |               | 令牌自检的认证方法名称。                                                                            |

Reviewer: ```suggestion
| introspection_endpoint_auth_method   | string  | 否     | "client_secret_basic" |               | 令牌自省的认证方法名称。                                                                            |
```

## 故障排除

1. 如果 APISIX 无法解析或者连接到身份认证服务（如 Okta、Keycloak、Authing 等），请检查或修改配置文件（`./conf/config.yaml`）中的 DNS 设置。

> 2. 如果遇到 `the error request to the redirect_uri path, but there's no session state found` 的错误，请确认当前访问的 URL 是否携带了 `code` 与 `state`，请勿直接访问 `redirect_uri`。

Reviewer 3: 如果遇到 `the error request to the redirect_uri path, but there's no session state found` 的错误，请检查 redirect_url 设置。 IdP 将请求重定向到一个预先配置好的URL（redirect_url），例如 `http://127.0.0.1:9080/callback`。请注意：这是一个在 APISIX 中不存在的 API，它只用于捕获相关的请求，并在 OIDC 逻辑中完成 Token 交换的功能。

Reviewer 3: `redirect_uri` is not been visited for you but for the IdP, so if this error happened, it means something wrong with IdP with APISIX.

Author: ![image](https://user-images.githubusercontent.com/2106987/178385376-c3a3ce31-b7e5-4d04-843c-4287f514ee98.png)

@tzssangglass @starsz Could you help add a suggestion directly?

> 2. 如果遇到 `the error request to the redirect_uri path, but there's no session state found` 的错误，请确认当前访问的 URL 是否携带了 `code` 与 `state`，请勿直接访问 `redirect_uri`。

Reviewer 3: ```suggestion
2. 如果遇到 `the error request to the redirect_uri path, but there's no session state found` 的错误，请检查 `redirect_url` 设置。APISIX 会向 IdP 发起身份认证请求，IdP 完成认证和授权后，会携带 ID Token和 Access Token 重定向到 `redirect_url` ，例如 `http://127.0.0.1:9080/callback`，此时再次进入 APISIX 并在 OIDC 逻辑中完成 Token 交换的功能。因此 `redirect_url` 需要满足以下条件：

-  `redirect_url`  需要能被当前 `openid-connect` 所在的路由捕获，比如当前路由的 `uri` 属性是 `/api/v1/*`, `redirect_uri` 可以是 `/api/v1/callback`
- `redirect_url` 的 scheme 和 host 是 IdP 视角下访问 APISIX 的属性。
```

and cc @starsz
